jpa/eclipselink.jpa.test.jse/src/org/eclipse/persistence/jpa/test/property/TestSQLCallDeferall.java

```diff
@@ -0,0 +1,58 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation. All rights reserved.
+ * This program and the accompanying materials are made available under the 
+ * terms of the Eclipse Public License v1.0 and Eclipse Distribution License v. 1.0 
+ * which accompanies this distribution. 
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at 
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * Contributors:
+ *     04/02/2018-2.7.0 Will Dazey
+ *       - 533148: Add SQL Call Deferral persistence property
+ ******************************************************************************/
+package org.eclipse.persistence.jpa.test.property;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+
+import org.eclipse.persistence.jpa.test.framework.DDLGen;
+import org.eclipse.persistence.jpa.test.framework.Emf;
+import org.eclipse.persistence.jpa.test.framework.EmfRunner;
+import org.eclipse.persistence.jpa.test.framework.Property;
+import org.eclipse.persistence.jpa.test.property.model.AbstractParent;
+import org.eclipse.persistence.jpa.test.property.model.Child;
+import org.eclipse.persistence.jpa.test.property.model.Parent;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(EmfRunner.class)
+public class TestSQLCallDeferall {
+
+    @Emf(createTables = DDLGen.DROP_CREATE, classes = { AbstractParent.class, Parent.class, Child.class }, properties = {
+            @Property(name = "eclipselink.jpa.sql-call-deferral", value = "false") })
+    private EntityManagerFactory emf;
+
+    /**
+     * Tests the 'eclipselink.jpa.sql-call-deferral' persistence property. Child entity sets a Foreign Key constraint
+     * that would lead EclipseLink to order SQL operations into a ForeignKey constraint violation. However, with
+     * the property set, the operations are ordered without deferral reordering.
+     */
+    @Test
+    public void testSQLDeferralProperty() {
+        EntityManager em = emf.createEntityManager();
+
+        em.getTransaction().begin();
+
+        Parent parent = new Parent();
+        em.persist(parent);
+
+        Child child = new Child();
+        child.setParent(parent);
+        em.persist(child);
+
+        // expects: INSERT order: AbstractParent, Parent, Child
+        // ECL default: INSERT order: AbstractParent, Child, Parent(deferred)
+        em.getTransaction().commit();
```

Comment:
I do not see any verification here. Are you looking for something like `org.eclipse.persistence.testing.framework.QuerySQLTracker` (there are more variants of this in tests)?

Reply:
The verification is that an SQL exception will be thrown on commit unless the property is set. If you look at the ```@JoinColumn``` on the Child entity, I am setting a custom foreignKey. Since EclipseLink doesn't process this FK, it would normally try to reorder INSERTs. However, with this new property, the reordering will not occur.

Reply:
I looked into using ```org.eclipse.persistence.sessions.SessionEventAdapter```, which is how ```org.eclipse.persistence.testing.framework.QuerySQLTracker``` seems to works, but there may be a bug in the listener code. If you look at ```AbstractSession.executeQuery(...)```, you can see the eventManager is logging the query execution. However, if you look at ```AbstractSession.executeCall()```, there is no eventManager calls. 

What this means is that only 2 queries are logged in the QuerySQLTracker: AbstractParent and Child. But the one call for Parent is not even logged. This may be a bug that needs to be fixed; I assumed ```SessionEventAdapter``` should log ALL sql, not just queries but calls as well. But in the mean time, I believe this test accurately fails when the property isnt set and passes if the property is deferring calls.